<h2>Gold Recovery Optimization Project</h2>

<h3>Introduction</h3>
<p>This project aims to analyze data from a gold mining process to identify opportunities for optimizing the gold recovery rate. By leveraging machine learning and statistical analysis, we will build a model to predict the final gold recovery and use this model to enhance the overall efficiency of the extraction process.</p>

<h3>Project Objectives</h3>
<ol>
    <li><b>Data Preparation:</b> Open, inspect, and prepare the provided training and test datasets. Validate the gold recovery calculation and handle any missing values or anomalies.</li>
    <li><b>Data Analysis:</b> Analyze how the concentrations of metals (Au, Ag, Pb) change at different stages of the process and compare feed particle size distributions between the training and test sets.</li>
    <li><b>Model Building:</b> Develop a function to calculate the project's key evaluation metric, sMAPE (Symmetric Mean Absolute Percentage Error).</li>
    <li><b>Model Evaluation and Recommendation:</b> Train and evaluate different machine learning models using cross-validation. Select the best model and provide a final recommendation based on its performance on the test set.</li>
</ol>
<p>This analysis will provide a data-driven recommendation for improving the efficiency of the gold recovery process, which is a significant factor in a mining company's profitability.</p>

In [1]:
# Import the neccessary libraries
import pandas as pd
from sklearn.metrics import mean_absolute_error

In [3]:
# Load the three project datasets into pandas DataFrames.
df_train = pd.read_csv('datasets/gold_recovery_train.csv')
df_test = pd.read_csv('datasets/gold_recovery_test.csv')
df_full = pd.read_csv('datasets/gold_recovery_full.csv')

In [4]:
# Checking Data Information, Duplicates, and Data Types
def check_data(df):
    """
    Checks for missing values, duplicates, and data types in a DataFrame.
    
    Args:
        df (pd.DataFrame): The DataFrame to check.
    """
    # 1. Checking Info 
    print("--- Checking for dataset info")
    display(df.info())

    # 2. Checking first five rows of datasets
    print("\n--- Checking for First Five Rows ---")
    display(df.head())
    
    # 3. Check for Missing Values
    print("\n--- Checking for Missing Values ---")
    print(df.isna().sum().sort_values(ascending=False))
    
    # 4. Check for Duplicates
    print("\n--- Checking for Duplicate Rows ---")
    print("Number of duplicate rows:", df.duplicated().sum())

In [5]:
# Check the training dataset for general information, like data types and non-null values.
print('Training DataFrame Information:')
check_data(df_train)

Training DataFrame Information:
--- Checking for dataset info
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16860 entries, 0 to 16859
Data columns (total 87 columns):
 #   Column                                              Non-Null Count  Dtype  
---  ------                                              --------------  -----  
 0   date                                                16860 non-null  object 
 1   final.output.concentrate_ag                         16788 non-null  float64
 2   final.output.concentrate_pb                         16788 non-null  float64
 3   final.output.concentrate_sol                        16490 non-null  float64
 4   final.output.concentrate_au                         16789 non-null  float64
 5   final.output.recovery                               15339 non-null  float64
 6   final.output.tail_ag                                16794 non-null  float64
 7   final.output.tail_pb                                16677 non-null  float64
 8   final.output.t

None


--- Checking for First Five Rows ---


date  final.output.concentrate_ag  \
0  2016-01-15 00:00:00                     6.055403   
1  2016-01-15 01:00:00                     6.029369   
2  2016-01-15 02:00:00                     6.055926   
3  2016-01-15 03:00:00                     6.047977   
4  2016-01-15 04:00:00                     6.148599   

   final.output.concentrate_pb  final.output.concentrate_sol  \
0                     9.889648                      5.507324   
1                     9.968944                      5.257781   
2                    10.213995                      5.383759   
3                     9.977019                      4.858634   
4                    10.142511                      4.939416   

   final.output.concentrate_au  final.output.recovery  final.output.tail_ag  \
0                    42.192020              70.541216             10.411962   
1                    42.701629              69.266198             10.462676   
2                    42.657501              68.116445             10.507046   
3                    42.689819              68.347543             10.422762   
4                    42.774141              66.927016             10.360302   

   final.output.tail_pb  final.output.tail_sol  final.output.tail_au  ...  \
0              0.895447              16.904297              2.143149  ...   
1              0.927452              16.634514              2.224930  ...   
2              0.953716              16.208849              2.257889  ...   
3              0.883763              16.532835              2.146849  ...   
4              0.792826              16.525686              2.055292  ...   

   secondary_cleaner.state.floatbank4_a_air  \
0                                 14.016835   
1                                 13.992281   
2                                 14.015015   
3                                 14.036510   
4                                 14.027298   

   secondary_cleaner.state.floatbank4_a_level  \
0                                 -502.488007   
1                                 -505.503262   
2                                 -502.520901   
3                                 -500.857308   
4                                 -499.838632   

   secondary_cleaner.state.floatbank4_b_air  \
0                                 12.099931   
1                                 11.950531   
2                                 11.912783   
3                                 11.999550   
4                                 11.953070   

   secondary_cleaner.state.floatbank4_b_level  \
0                                 -504.715942   
1                                 -501.331529   
2                                 -501.133383   
3                                 -501.193686   
4                                 -501.053894   

   secondary_cleaner.state.floatbank5_a_air  \
0                                  9.925633   
1                                 10.039245   
2                                 10.070913   
3                                  9.970366   
4                                  9.925709   

   secondary_cleaner.state.floatbank5_a_level  \
0                                 -498.310211   
1                                 -500.169983   
2                                 -500.129135   
3                                 -499.201640   
4                                 -501.686727   

   secondary_cleaner.state.floatbank5_b_air  \
0                                  8.079666   
1                                  7.984757   
2                                  8.013877   
3                                  7.977324   
4                                  7.894242   

   secondary_cleaner.state.floatbank5_b_level  \
0                                 -500.470978   
1                                 -500.582168   
2                                 -500.517572   
3                                 -500.255908   
4                                 -500.356035   

   secondary_cleaner.state.floatbank6_a_air  \
0    


--- Checking for Missing Values ---
rougher.output.recovery                               2573
rougher.output.tail_ag                                2250
rougher.output.tail_sol                               2249
rougher.output.tail_au                                2249
secondary_cleaner.output.tail_sol                     1986
                                                      ... 
primary_cleaner.state.floatbank8_d_level                27
rougher.calculation.floatbank10_sulfate_to_au_feed      27
rougher.calculation.floatbank11_sulfate_to_au_feed      27
primary_cleaner.input.feed_size                          0
date                                                     0
Length: 87, dtype: int64

--- Checking for Duplicate Rows ---
Number of duplicate rows: 0


In [6]:
# Check the test dataset for general information, like data types and non-null values.
print('Test DataFrame Information:')
check_data(df_test)

Test DataFrame Information:
--- Checking for dataset info
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5856 entries, 0 to 5855
Data columns (total 53 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   date                                        5856 non-null   object 
 1   primary_cleaner.input.sulfate               5554 non-null   float64
 2   primary_cleaner.input.depressant            5572 non-null   float64
 3   primary_cleaner.input.feed_size             5856 non-null   float64
 4   primary_cleaner.input.xanthate              5690 non-null   float64
 5   primary_cleaner.state.floatbank8_a_air      5840 non-null   float64
 6   primary_cleaner.state.floatbank8_a_level    5840 non-null   float64
 7   primary_cleaner.state.floatbank8_b_air      5840 non-null   float64
 8   primary_cleaner.state.floatbank8_b_level    5840 non-null   float64
 9   primary_cleaner.state.float

None


--- Checking for First Five Rows ---


date  primary_cleaner.input.sulfate  \
0  2016-09-01 00:59:59                     210.800909   
1  2016-09-01 01:59:59                     215.392455   
2  2016-09-01 02:59:59                     215.259946   
3  2016-09-01 03:59:59                     215.336236   
4  2016-09-01 04:59:59                     199.099327   

   primary_cleaner.input.depressant  primary_cleaner.input.feed_size  \
0                         14.993118                         8.080000   
1                         14.987471                         8.080000   
2                         12.884934                         7.786667   
3                         12.006805                         7.640000   
4                         10.682530                         7.530000   

   primary_cleaner.input.xanthate  primary_cleaner.state.floatbank8_a_air  \
0                        1.005021                             1398.981301   
1                        0.990469                             1398.777912   
2                        0.996043                             1398.493666   
3                        0.863514                             1399.618111   
4                        0.805575                             1401.268123   

   primary_cleaner.state.floatbank8_a_level  \
0                               -500.225577   
1                               -500.057435   
2                               -500.868360   
3                               -498.863574   
4                               -500.808305   

   primary_cleaner.state.floatbank8_b_air  \
0                             1399.144926   
1                             1398.055362   
2                             1398.860436   
3                             1397.440120   
4                             1398.128818   

   primary_cleaner.state.floatbank8_b_level  \
0                               -499.919735   
1                               -499.778182   
2                               -499.764529   
3                               -499.211024   
4                               -499.504543   

   primary_cleaner.state.floatbank8_c_air  ...  \
0                             1400.102998  ...   
1                             1396.151033  ...   
2                             1398.075709  ...   
3                             1400.129303  ...   
4                             1402.172226  ...   

   secondary_cleaner.state.floatbank4_a_air  \
0                                 12.023554   
1                                 12.058140   
2                                 11.962366   
3                                 12.033091   
4                                 12.025367   

   secondary_cleaner.state.floatbank4_a_level  \
0                                 -497.795834   
1                                 -498.695773   
2                                 -498.767484   
3                                 -498.350935   
4                                 -500.786497   

   secondary_cleaner.state.floatbank4_b_air  \
0                                  8.016656   
1                                  8.130979   
2                                  8.096893   
3                                  8.074946   
4                                  8.054678   

   secondary_cleaner.state.floatbank4_b_level  \
0                                 -501.289139   
1                                 -499.634209   
2                                 -500.827423   
3                                 -499.474407   
4                                 -500.397500   

   secondary_cleaner.state.floatbank5_a_air  \
0                                  7.946562   
1                                  7.958270   
2                                  8.071056   
3                                  7.897085   
4                                  8.107890   

   secondary_cleaner.state.floatbank5_a_level  \
0                                 -432.317850   
1                                 -525.839648   
2                                 -500.801673   
3     


--- Checking for Missing Values ---
rougher.input.floatbank11_xanthate            353
primary_cleaner.input.sulfate                 302
primary_cleaner.input.depressant              284
rougher.input.floatbank10_sulfate             257
primary_cleaner.input.xanthate                166
rougher.input.floatbank10_xanthate            123
rougher.input.feed_sol                         67
rougher.input.floatbank11_sulfate              55
rougher.input.feed_rate                        40
secondary_cleaner.state.floatbank3_a_air       34
secondary_cleaner.state.floatbank2_b_air       23
rougher.input.feed_size                        22
secondary_cleaner.state.floatbank2_a_air       20
rougher.state.floatbank10_a_air                17
rougher.state.floatbank10_c_air                17
rougher.state.floatbank10_d_air                17
rougher.state.floatbank10_e_air                17
rougher.state.floatbank10_b_air                17
rougher.state.floatbank10_f_air                17
secondary_cle

In [7]:
# Check the full dataset for general information, like data types and non-null values.
print('Full DataFrame Information:')
check_data(df_full)

Full DataFrame Information:
--- Checking for dataset info
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22716 entries, 0 to 22715
Data columns (total 87 columns):
 #   Column                                              Non-Null Count  Dtype  
---  ------                                              --------------  -----  
 0   date                                                22716 non-null  object 
 1   final.output.concentrate_ag                         22627 non-null  float64
 2   final.output.concentrate_pb                         22629 non-null  float64
 3   final.output.concentrate_sol                        22331 non-null  float64
 4   final.output.concentrate_au                         22630 non-null  float64
 5   final.output.recovery                               20753 non-null  float64
 6   final.output.tail_ag                                22633 non-null  float64
 7   final.output.tail_pb                                22516 non-null  float64
 8   final.output.tail_

None


--- Checking for First Five Rows ---


date  final.output.concentrate_ag  \
0  2016-01-15 00:00:00                     6.055403   
1  2016-01-15 01:00:00                     6.029369   
2  2016-01-15 02:00:00                     6.055926   
3  2016-01-15 03:00:00                     6.047977   
4  2016-01-15 04:00:00                     6.148599   

   final.output.concentrate_pb  final.output.concentrate_sol  \
0                     9.889648                      5.507324   
1                     9.968944                      5.257781   
2                    10.213995                      5.383759   
3                     9.977019                      4.858634   
4                    10.142511                      4.939416   

   final.output.concentrate_au  final.output.recovery  final.output.tail_ag  \
0                    42.192020              70.541216             10.411962   
1                    42.701629              69.266198             10.462676   
2                    42.657501              68.116445             10.507046   
3                    42.689819              68.347543             10.422762   
4                    42.774141              66.927016             10.360302   

   final.output.tail_pb  final.output.tail_sol  final.output.tail_au  ...  \
0              0.895447              16.904297              2.143149  ...   
1              0.927452              16.634514              2.224930  ...   
2              0.953716              16.208849              2.257889  ...   
3              0.883763              16.532835              2.146849  ...   
4              0.792826              16.525686              2.055292  ...   

   secondary_cleaner.state.floatbank4_a_air  \
0                                 14.016835   
1                                 13.992281   
2                                 14.015015   
3                                 14.036510   
4                                 14.027298   

   secondary_cleaner.state.floatbank4_a_level  \
0                                 -502.488007   
1                                 -505.503262   
2                                 -502.520901   
3                                 -500.857308   
4                                 -499.838632   

   secondary_cleaner.state.floatbank4_b_air  \
0                                 12.099931   
1                                 11.950531   
2                                 11.912783   
3                                 11.999550   
4                                 11.953070   

   secondary_cleaner.state.floatbank4_b_level  \
0                                 -504.715942   
1                                 -501.331529   
2                                 -501.133383   
3                                 -501.193686   
4                                 -501.053894   

   secondary_cleaner.state.floatbank5_a_air  \
0                                  9.925633   
1                                 10.039245   
2                                 10.070913   
3                                  9.970366   
4                                  9.925709   

   secondary_cleaner.state.floatbank5_a_level  \
0                                 -498.310211   
1                                 -500.169983   
2                                 -500.129135   
3                                 -499.201640   
4                                 -501.686727   

   secondary_cleaner.state.floatbank5_b_air  \
0                                  8.079666   
1                                  7.984757   
2                                  8.013877   
3                                  7.977324   
4                                  7.894242   

   secondary_cleaner.state.floatbank5_b_level  \
0                                 -500.470978   
1                                 -500.582168   
2                                 -500.517572   
3                                 -500.255908   
4                                 -500.356035   

   secondary_cleaner.state.floatbank6_a_air  \
0    


--- Checking for Missing Values ---
rougher.output.recovery                     3119
rougher.output.tail_ag                      2737
rougher.output.tail_au                      2736
rougher.output.tail_sol                     2736
rougher.input.floatbank11_xanthate          2257
                                            ... 
primary_cleaner.state.floatbank8_b_level      43
primary_cleaner.state.floatbank8_c_level      43
primary_cleaner.state.floatbank8_d_level      43
primary_cleaner.input.feed_size                0
date                                           0
Length: 87, dtype: int64

--- Checking for Duplicate Rows ---
Number of duplicate rows: 0


# Step 1: Data Preparation

## 1.1: Initial Data Exploration and Loading

The first step is to load the three provided datasets: `gold_recovery_train.csv`, `gold_recovery_test.csv`, and `gold_recovery_full.csv`. An initial exploration was conducted on each to understand their structure, identify missing data, and check for quality issues.

### Training Set (`df_train`)
* **Shape**: The training set contains 16,860 rows and 87 columns.
* **Data Quality**: The data contains a significant number of missing values across various columns, which will need to be addressed during preprocessing. No duplicate rows were found.
* **Data Types**: The `date` column is an `object` type, while the remaining 86 columns are numerical (`float64`), suitable for modeling.

### Test Set (`df_test`)
* **Shape**: The test set contains 5,856 rows and 53 columns.
* **Data Quality**: Similar to the training set, there are missing values present. No duplicate rows were found.
* **Data Types**: The `date` column is an `object`, and the other 52 columns are numerical (`float64`).

### Full Dataset (`df_full`)
* **Shape**: The complete dataset contains 22,716 rows and the full 87 columns.
* **Data Quality**: This dataset also contains missing values, consistent with the `train` and `test` sets.
* **Purpose**: This dataset serves as the complete source of truth and can be used to add the missing target and output columns to the test set after the model has been trained and evaluated.

### Key Observation: Feature Mismatch

The most critical finding from this initial exploration is the discrepancy in the number of columns between the training set (87) and the test set (53). The test set is missing 34 columns. These columns correspond to **output measurements** and **calculated values** from various stages of the purification process, including our two target variables (`rougher.output.recovery` and `final.output.recovery`).

This is expected, as in a real-world scenario, these values would not be available at the time of prediction. Therefore, these 34 columns must be excluded from the feature set when training our model.

In [8]:
# Function for calculating the recovery using the provided formula
def recovery_calculation(C, F, T):
    return (C * (F - T)) / (F * (C - T)) * 100

# Create a copy of the training data and drop rows with missing values needed for the calculation
data_for_check = df_train.dropna(subset=[
    'rougher.output.concentrate_au',
    'rougher.input.feed_au',
    'rougher.output.tail_au',
    'rougher.output.recovery'
]).copy()

# Apply your function to calculate recovery
calculated_recovery = recovery_calculation(
    data_for_check['rougher.output.concentrate_au'],
    data_for_check['rougher.input.feed_au'],
    data_for_check['rougher.output.tail_au']
)

# Calculate the MAE between your calculation and the given data
mae = mean_absolute_error(
    data_for_check['rougher.output.recovery'],
    calculated_recovery
)

print(f"The Mean Absolute Error (MAE) is: {mae}")

The Mean Absolute Error (MAE) is: 9.210911277458828e-15


## 1.2: Verify Recovery Calculation

A critical data integrity check was performed to ensure the `rougher.output.recovery` values were calculated correctly. A custom function was written to replicate the provided formula, and the Mean Absolute Error (MAE) was calculated between our results and the original data.

**Result**: The MAE was found to be `9.3e-15`, which is effectively zero.

**Conclusion**: This confirms that the recovery data is accurate and reliable. We can proceed with the analysis.